pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,522 @@
+		nodeSelectors = append(nodeSelectors, &ipPool.Spec.NodeSelector)
+	}
+
+	return
+}
+
+func (c *Cluster) filterNodes() (nodes []*corev1.Node, err error) {
+	nodeSelectors, err := c.listNodeSelectors()
+	if err != nil {
+		klog.ErrorS(err, "List node selectors of nodes failed")
+		return
+	}
+	nodeSet := sets.String{}
+	for _, nodeSelector := range nodeSelectors {
+		nodeSel, _ := metav1.LabelSelectorAsSelector(nodeSelector)
+		nodeList, _ := c.nodeLister.List(nodeSel)
+		for _, node := range nodeList {
+			nodeName := node.Name
+			if !nodeSet.Has(nodeName) {
+				nodeSet.Insert(node.Name)
+				nodes = append(nodes, node)
+			}
+		}
+	}
+	return
+}
+
+func (c *Cluster) syncClusterMembers() (clusterNodes []string) {
+	nodes, err := c.filterNodes()
+	if err != nil {
+		klog.ErrorS(err, "Listing Nodes failed")
+		return
+	}
+	klog.V(3).InfoS("Sync cluster members", "node num", len(nodes))
+
+	for _, node := range nodes {
+		nodeAddr, err := k8s.GetNodeAddr(node)
+		if err != nil {
+			klog.ErrorS(err, "Obtain local IP address from K8s node failed")
+			continue
+		}
+		member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+		clusterNodes = append(clusterNodes, member)
+	}
+	return
+}
+
+func (c *Cluster) shouldJoinCluster(nodeName string) bool {
+	node, err := c.nodeLister.Get(nodeName)
+	if err != nil {
+		klog.ErrorS(err, "Get node failed", "nodeName", nodeName)
+		return false
+	}
+	return c.matchNodeSelectors(node)
+}
+
+func (c *Cluster) matchNodeSelectors(node *corev1.Node) bool {
+	nodeSelectors, err := c.listNodeSelectors()
+	if err != nil {
+		klog.ErrorS(err, "matchNodeSelectors failed")
+		return false
+	}
+	if nodeSelectors == nil {
+		return true
+	}
+	for _, ns := range nodeSelectors {
+		nodeSelector, _ := metav1.LabelSelectorAsSelector(ns)
+		if ns == nil {
+			return true
+		}
+		if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *Cluster) memberNum() int {
+	c.memberRWLock.RLock()
+	defer c.memberRWLock.RUnlock()
+	return c.existingMembers.Len()
+}
+
+func (c *Cluster) addExistingMembers(members ...string) {
+	c.memberRWLock.Lock()
+	defer c.memberRWLock.Unlock()
+	c.existingMembers.Insert(members...)
+	klog.V(2).InfoS("Add existingMember", "memberNum", c.existingMembers.Len())
+}
+
+func (c *Cluster) updateExistingMember(node *corev1.Node, add bool) {
+	c.memberRWLock.Lock()
+	defer c.memberRWLock.Unlock()
+	nodeAddr, err := k8s.GetNodeAddr(node)
+	if err != nil {
+		klog.ErrorS(err, "Failed to obtain local IP address")
+		return
+	}
+	member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+	oldNum := c.existingMembers.Len()
+	if add {
+		c.existingMembers.Insert(member)
+		newNum := c.existingMembers.Len()
+		klog.V(2).Infof("Add new member (%s) in local existing memberlist, num %d->%d", member, oldNum, newNum)
+	} else {
+		if !c.existingMembers.Has(member) {
+			return
+		}
+		c.existingMembers.Delete(member)
+		newNum := c.existingMembers.Len()
```

Comment:
Doesn't it need to rejoin the new node when the list changes? Or it waits for other Nodes join this node?

Reply:
Only local node which has not join cluster need to join, the joined node knows cluster members through memberlist.
I think there is no need to join here.